sql/sql_admission_control.h

```diff
@@ -0,0 +1,356 @@
+  ADMISSION_CONTROL_ALTER,
+  ADMISSION_CONTROL_BEGIN,
+  ADMISSION_CONTROL_COMMIT,
+  ADMISSION_CONTROL_CREATE,
+  ADMISSION_CONTROL_DELETE,
+  ADMISSION_CONTROL_DROP,
+  ADMISSION_CONTROL_INSERT,
+  ADMISSION_CONTROL_LOAD,
+  ADMISSION_CONTROL_SELECT,
+  ADMISSION_CONTROL_SET,
+  ADMISSION_CONTROL_REPLACE,
+  ADMISSION_CONTROL_ROLLBACK,
+  ADMISSION_CONTROL_TRUNCATE,
+  ADMISSION_CONTROL_UPDATE,
+  ADMISSION_CONTROL_SHOW,
+  ADMISSION_CONTROL_USE,
+  ADMISSION_CONTROL_END = 64
+};
+
+/*
+ * sql_admission_control.h/cc
+ *
+ * This module handles multi-tenancy resource allocation on the server side.
+ * Functions will call into multi-tenancy plugin interfaces (if installed) to
+ * make a decision of whether or not a resource can be allocated. Connections
+ * and query limits are among the resources multi-tenancy plugin allocates.
+ * Note that admission control is now part of the multi-tenancy module to
+ * hanlde query throttling.
+ *
+ * The isolation level is defined by an entity. The entity could be a database,
+ * a user, or any thing that multi-tenancy plugin defines to isolate the
+ * resource allocation.
+ *
+ * See sql_admission_control.cc for implementation.
+ */
+
+extern int multi_tenancy_admit_query(THD *, const MT_RESOURCE_ATTRS *);
+extern int multi_tenancy_exit_query(THD *, const MT_RESOURCE_ATTRS *);
+
+/**
+  Per-thread information used in admission control.
+*/
+
+#ifdef HAVE_PSI_INTERFACE
+
+static PSI_mutex_key key_ac_node_lock;
+static PSI_cond_key key_ac_node_cond;
+
+static PSI_mutex_info all_ac_node_mutexes[] = {
+    {&key_ac_node_lock, "st_ac_node::lock", PSI_FLAG_SINGLETON, 0,
+     PSI_DOCUMENT_ME}};
+
+static PSI_cond_info all_ac_node_conds[] = {
+    {&key_ac_node_cond, "st_ac_node::cond", PSI_FLAG_SINGLETON, 0,
+     PSI_DOCUMENT_ME}};
+
+static void init_st_ac_node_keys() {
+  const char *const category = "sql";
+  int count;
+
+  count = static_cast<int>(array_elements(all_ac_node_mutexes));
+  mysql_mutex_register(category, all_ac_node_mutexes, count);
+
+  count = static_cast<int>(array_elements(all_ac_node_conds));
+  mysql_cond_register(category, all_ac_node_conds, count);
+}
+#endif
+
+struct st_ac_node {
+  mysql_mutex_t lock;
+  mysql_cond_t cond;
+  st_ac_node() {
+#ifdef HAVE_PSI_INTERFACE
+    init_st_ac_node_keys();
+#endif
+    mysql_mutex_init(key_ac_node_lock, &lock, MY_MUTEX_INIT_FAST);
+    mysql_cond_init(key_ac_node_cond, &cond);
+  }
+
+  ~st_ac_node() {
+    mysql_mutex_destroy(&lock);
+    mysql_cond_destroy(&cond);
+  }
+};
+
+/**
+  Class used in admission control.
+
+  Every entity (database or table or user name) will have this
+  object created and stored in the global map AC::ac_map.
+
+*/
+class Ac_info {
+  friend class AC;
+#ifdef HAVE_PSI_INTERFACE
+  PSI_mutex_key key_lock;
+  PSI_mutex_info key_lock_info[1] = {{&key_lock, "Ac_info::lock", 0}};
+#endif
+  // Queue to track the running and waiting threads.
+  std::deque<std::shared_ptr<st_ac_node>> queue;
```

Comment:
usings/typedef for custom containers
```suggestion
  using ac_node_ptr_container = std::deque<st_ac_node_ptr>;
  ac_node_ptr_container queue;
```